src/contracts/Output/VisitorAdapterNormalizer.php

```diff
@@ -0,0 +1,160 @@
+            return $this->visitValueObject($object, $eligibleVisitor, $format, $context);
+        }
+
+        return $this->normalizer->normalize($object, $format, $context);
+    }
+
+    /**
+     * @param array<string, mixed> $context
+     */
+    public function supportsNormalization(mixed $data, ?string $format = null, array $context = []): bool
+    {
+        if (($context[self::CALLED_CONTEXT] ?? false) === true) {
+            return false;
+        }
+
+        $eligibleVisitor = is_object($data)
+            ? $this->valueObjectVisitorResolver->resolveValueObjectVisitor($data)
+            : null;
+
+        if ($eligibleVisitor instanceof ValueObjectVisitor) {
+            return true;
+        }
+
+        if (!$this->normalizer instanceof ContextAwareNormalizerInterface) {
+            throw new LogicException(
+                sprintf(
+                    'Normalizer "%s" must be an instance of "%s".',
+                    $this->normalizer::class,
+                    ContextAwareNormalizerInterface::class,
+                ),
+            );
+        }
+
+        return $this->normalizer->supportsNormalization(
+            $data,
+            $format,
+            $context + [self::CALLED_CONTEXT => true],
+        );
+    }
+
+    /**
+     * @param array<mixed> $context
+     *
+     * @return array<mixed>
+     */
+    private function visitValueObject(
+        object $object,
+        ValueObjectVisitor $valueObjectVisitor,
+        ?string $format,
+        array $context,
+    ): array {
+        $visitor = $context['visitor'] ?? $this->createVisitor($format);
+        $generator = $visitor->getGenerator();
+
+        $generator->reset();
+        $generator->startDocument($object);
+
+        $valueObjectVisitor->visit($visitor, $generator, $object);
+
+        $generator->endDocument($object);
+
+        $data = $generator->getData();
+
+        $normalizedData = $this->normalizer->normalize(
+            $data,
+            $format,
+            $this->buildContext($context, $format),
+        );
+
+        return $normalizedData + [self::ENCODER_CONTEXT => $generator->getEncoderContext(get_object_vars($data))];
+    }
+
+    /**
+     * @param array<mixed> $context
+     *
+     * @return array<mixed>
+     */
+    private function buildContext(array $context, ?string $format): array
+    {
+        $context += [self::CALLED_CONTEXT => true];
+
+        if ($format === 'xml') {
+            $context += [Generator\Xml::OUTER_ELEMENT => true];
+        }
+
+        return $context;
+    }
+
+    private function createGenerator(string $format): BaseGenerator
+    {
+        if ($format === 'xml') {
+            return new Generator\Xml(
+                new Generator\Xml\FieldTypeHashGenerator($this->normalizer),
+            );
+        }
+
+        return new Generator\Json(
+            new Generator\Json\FieldTypeHashGenerator($this->normalizer),
+        );
```

Comment:
What if format is going to get something else than `xml` or `json`? I'd rather expect here either `InvalidArgumentException` or `LogicException` if that's not possible due to method call flow.